multi_gpu/four/model_parallel.py

```diff
@@ -0,0 +1,162 @@
+        self.conv1 = nn.Conv2d(in_channels=3, out_channels=48, kernel_size=(3,3), padding=(1,1))
+        self.conv1 = nn.DataParallel(self.conv1)
+        self.conv2 = nn.Conv2d(in_channels=48, out_channels=96, kernel_size=(3,3), padding=(1,1))
+        self.conv2 = nn.DataParallel(self.conv2)
+
+        self.conv3 = nn.Conv2d(in_channels=96, out_channels=192, kernel_size=(3,3), padding=(1,1))
+        self.conv3 = nn.DataParallel(self.conv3)
+
+        self.conv4 = nn.Conv2d(in_channels=192, out_channels=256, kernel_size=(3,3), padding=(1,1))
+        self.conv4 = nn.DataParallel(self.conv4)
+        
+        self.conv5 = nn.Conv2d(in_channels=256, out_channels=512, kernel_size=(3,3), padding=(1,1))
+        self.conv5 = nn.DataParallel(self.conv5)        
+
+        self.conv6 = nn.Conv2d(in_channels=512, out_channels=1024, kernel_size=(3,3), padding=(1,1))
+        self.conv6 = nn.DataParallel(self.conv6)
+
+        self.pool = nn.MaxPool2d(2,2)
+        self.fc1 = nn.Linear(in_features=8*8*1024, out_features=2048)
+        self.fc1 = nn.DataParallel(self.fc1)
+
+        self.fc2 = nn.Linear(in_features=2048, out_features=64)
+        self.fc2 = nn.DataParallel(self.fc2)
+
+        self.Dropout = nn.Dropout(0.25)
+        self.fc3 = nn.Linear(in_features=64, out_features=10)
+        self.fc3 = nn.DataParallel(self.fc3)
+
+    def forward(self, x):
+        x = F.relu(self.conv1(x)) #32*32*48
+        x = F.relu(self.conv2(x)) #32*32*96
+        x = self.pool(x) #16*16*96
+        x = self.Dropout(x)
+        x = F.relu(self.conv3(x)) #16*16*192
+        x = F.relu(self.conv4(x)) #16*16*256
+        x = F.relu(self.conv5(x)) #16*16*256
+        x = F.relu(self.conv6(x)) #16*16*256
+
+        x = self.pool(x) # 8*8*256
+        x = self.Dropout(x)
+        x = x.view(-1, 8*8*1024) # reshape x
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.Dropout(x)
+        x = self.fc3(x)
+        return x
+# Define loss function and optimizer. We employ cross-entropy and Adam
+import torch.optim as optim
+
+criterion = nn.CrossEntropyLoss()
+# define the test accuracy function
+def test_accuracy(net, testset_loader, epoch):
+    # Test the model
+    net.eval()
+    correct = 0
+    total = 0
+    for data in testset_loader:
+        images, labels = data
+        images, labels = Variable(images).cuda(), labels.cuda()
+        output = net(images)
+        _, predicted = torch.max(output.data, 1)
+        total += labels.size(0)
+        correct += (predicted == labels).sum()
+    print('Accuracy of the network after epoch '+str(epoch+1)+' is: ' + str(100 * correct / total))
+
+#We save the model after every 5 epochs
+def save_model(net, epoch):
+    filename = "model_with_epoch" + str(epoch+1) + ".pth"
+    torch.save(net.state_dict(),filename)
+
+#Train the neural network and test for accuracy after every 5 epochs
+#Depend on whether we need to load the pretrained model
+
+net = ConvNet()
```

Comment:
instead of using data parallel on single layers can just do
`net= ConvNet()`
`net = nn.DataParallel(net, device_ids=[0,1])`
if you wanted to train on 2 gpus on the same machine. If you don't include device_ids it will defaults to all available gpus on a machine.

Reply:
@b-kaufman do you know what would happen if you use `nn.DataParallel` when the machine has a single GPU available?  Does that degrade performance, or is PyTorch smart enough to recognize there isn't any parallelism needed?

I'm thinking about how to simplify the code and whether a single `.py` file can trivially run the 1, 2, and 4 GPU cases without arguments.  HTCondor will control the number of available GPUs on the server via `$CUDA_VISIBLE_DEVICES`, which is what `nn.DataParallel` will see for the default GPU devices.  If we don't specify the GPUs, we might automatically get 1, 2, or 4.

Reply:
@agitter if you don't specify device Ids it should use all visible devices. You could use the same code for all three cases but I'm not certain about degradation. Again though the best practice for multiple GPUs would be DistributedDataParallel. This is even recommended in the docs for DataParallel: https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html